Assets/Mirror/Components/InterestManagement/Match/MatchInterestManagement.cs

```diff
@@ -7,129 +7,123 @@ namespace Mirror
-            // Multiple objects could be destroyed in same frame and we don't
-            // want to rebuild for each one...let Update do it once.
-            // We must add the current match to dirtyMatches for Update to rebuild it.
-            if (lastObjectMatch.TryGetValue(identity, out Guid currentMatch))
-            {
-                lastObjectMatch.Remove(identity);
-                if (currentMatch != Guid.Empty && matchObjects.TryGetValue(currentMatch, out HashSet<NetworkIdentity> objects) && objects.Remove(identity))
-                    dirtyMatches.Add(currentMatch);
-            }
+            foreach (NetworkMatch networkMatch in matchObjects[matchId])
+                if (networkMatch.netIdentity != null)
+                    NetworkServer.RebuildObservers(networkMatch.netIdentity, false);
         }
 
-        // internal so we can update from tests
+        // called by NetworkMatch.matchId setter
         [ServerCallback]
-        internal void Update()
+        internal void OnMatchChanged(NetworkMatch networkMatch)
         {
-            // for each spawned:
-            //   if match changed:
-            //     add previous to dirty
-            //     add new to dirty
-            foreach (NetworkIdentity identity in NetworkServer.spawned.Values)
-            {
-                // Ignore objects that don't have a NetworkMatch component
-                if (!identity.TryGetComponent(out NetworkMatch networkMatch))
-                    continue;
-
-                Guid newMatch = networkMatch.matchId;
-                if (!lastObjectMatch.TryGetValue(identity, out Guid currentMatch))
-                    continue;
-
-                // Guid.Empty is never a valid matchId
-                // Nothing to do if matchId hasn't changed
-                if (newMatch == Guid.Empty || newMatch == currentMatch)
-                    continue;
-
-                // Mark new/old matches as dirty so they get rebuilt
-                UpdateDirtyMatches(newMatch, currentMatch);
+            // Mark new/old matches as dirty so they get rebuilt
+            UpdateDirtyMatches(networkMatch.matchId, networkMatch);
 
+            if (lastObjectMatch.TryGetValue(networkMatch.netIdentity, out NetworkMatch currentMatch))
+            {
                 // This object is in a new match so observers in the prior match
                 // and the new match need to rebuild their respective observers lists.
-                UpdateMatchObjects(identity, newMatch, currentMatch);
+                UpdateMatchObjects(networkMatch.netIdentity, networkMatch, currentMatch);
             }
-
-            // rebuild all dirty matches
-            foreach (Guid dirtyMatch in dirtyMatches)
-                RebuildMatchObservers(dirtyMatch);
-
-            dirtyMatches.Clear();
         }
 
-        void UpdateDirtyMatches(Guid newMatch, Guid currentMatch)
+        [ServerCallback]
+        void UpdateDirtyMatches(Guid newMatch, NetworkMatch currentMatch)
```

Comment:
Only one usage and a fairly small method, could be inlined above?

Reply:
Merged them together